### PR TITLE
M 003 create enum interface

### PR DESCRIPTION
**Pull request by YaroslavKlochko**

### CODE REVIEW

File: src/main/java/com/mhcb/domain/Form.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+@Data
+@Entity
+@Table(name = "forms")
+@AllArgsConstructor
+@NoArgsConstructor
+public class Form {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/Form.java]: Could you please change strategy to SEQUENCE?

File: src/main/java/com/mhcb/domain/FormConverter.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.mhcb.domain;
+
+import javax.persistence.AttributeConverter;
+import javax.persistence.Converter;
+
+@Converter(autoApply = true)
+public class FormConverter implements AttributeConverter<FormState, String> {
+
+    @Override
+    public String convertToDatabaseColumn(FormState state) {
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/FormConverter.java]: please add `final`

File: src/main/java/com/mhcb/domain/FormConverter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        if (state == null) {
+            return "";
+        }
+            return state.getCurrentState();
+    }
+
+    @Override
+    public FormState convertToEntityAttribute(String state) {
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/FormConverter.java]: please add `final`

File: src/main/java/com/mhcb/domain/FormStateMicroFactory.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.mhcb.domain;
+
+import static java.util.Objects.isNull;
+
+public class FormStateMicroFactory {
+
+    public FormStateMicro getFormStateMicro(String stateType) {
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/FormStateMicroFactory.java]: please add `final`

File: src/main/java/com/mhcb/domain/FormStateMicroFactory.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+        if (isNull(stateType)) {
+            throw new IllegalArgumentException("State type cannot be null.");
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/FormStateMicroFactory.java:9]: Could we catch this exception? When? Where? How do we deal with it?

File: src/main/java/com/mhcb/domain/State.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    NEW, VETTING, REJECT_VETTING, PHONE_INTERVIEW, REJECT_PHONE_INTERVIEW, FIELD_VISIT, REJECT_FIELD_VISIT,
+    WAITING_FOR_APPROVAL, APPROVED, REJECT_APPROVED, DOCUMENTS, WAITING_SIGNING, SIGNED, FUNDED,
+    WAITING_RECEIPTS, RECEIPTS_SUBMITTED, REJECT_FOREVER, MONITORING, MONITORING_COMPLETED, PENDING;
+
+    private String stateType;
+
+    State() {
+    }
+
+    private State(String stateType) {
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/State.java]: please add `final`

**Review by sergiiManko — CHANGES_REQUESTED** [review]

File: src/main/java/com/mhcb/domain/FormConverter.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import javax.persistence.AttributeConverter;
+import javax.persistence.Converter;
+
+import static java.util.Objects.isNull;
+
+@Converter(autoApply = true)
+public class FormConverter implements AttributeConverter<FormState, String> {
+
+    @Override
+    public String convertToDatabaseColumn(FormState state) {
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/domain/FormConverter.java]: Could you please add the `final` keyword?

File: src/main/java/com/mhcb/exception/FormStateExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+@ControllerAdvice
+public class FormStateExceptionHandler {
+
+    @ExceptionHandler(FormStateRequestException.class)
+    public ResponseEntity<Object> handleFormStateRequestException(final FormStateRequestException e) {
+        final HttpStatus badRequest = HttpStatus.BAD_REQUEST;
+        final FormStateException formStateException = new FormStateException
+                (
+                e.getMessage(),
+                ZonedDateTime.now(ZoneId.of("Z"))
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/exception/FormStateExceptionHandler.java:20]: Do we need to use ZonedDateTime?

File: src/main/java/com/mhcb/exception/FormStateExceptionHandler.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+
+@ControllerAdvice
+public class FormStateExceptionHandler {
+
+    @ExceptionHandler(FormStateRequestException.class)
+    public ResponseEntity<Object> handleFormStateRequestException(final FormStateRequestException e) {
+        final HttpStatus badRequest = HttpStatus.BAD_REQUEST;
```
Comment (sergiiManko) [inline src/main/java/com/mhcb/exception/FormStateExceptionHandler.java:16]: Why do you decide to extract HTTP status to a separate variable?

**Review by sergiiManko — CHANGES_REQUESTED** [review]

File: src/main/java/com/mhcb/domain/FormConverter.java (hunk 7 above)
Comment (YaroslavKlochko) [reply src/main/java/com/mhcb/domain/FormConverter.java]: done

File: src/main/java/com/mhcb/exception/FormStateExceptionHandler.java (hunk 8 above)
Comment (YaroslavKlochko) [reply src/main/java/com/mhcb/exception/FormStateExceptionHandler.java:20]: done

File: src/main/java/com/mhcb/exception/FormStateExceptionHandler.java (hunk 9 above)
Comment (YaroslavKlochko) [reply src/main/java/com/mhcb/exception/FormStateExceptionHandler.java:16]: done

**Pull request closed without merging**